packages/grafana-runtime/src/services/AggregateHistory.ts

```diff
@@ -0,0 +1,183 @@
+ * @param param Name of the query param for secondary location
+ * @param isMain
+ */
+function getPartialLocation(location: H.Location, param: string, isMain: boolean): H.Location {
+  const newLocation = { ...location };
+  const paramsCopy = new URLSearchParams(newLocation.search);
+  if (isMain) {
+    paramsCopy.delete(param);
+    newLocation.search = paramsCopy.toString();
+    return newLocation;
+  } else {
+    if (paramsCopy.has(param)) {
+      const url = paramsCopy.get(param);
+      const parsed = new URL('http://sidecar' + url);
+      return {
+        pathname: parsed.pathname,
+        search: parsed.search,
+        hash: parsed.hash,
+        state: undefined,
+      };
+    } else {
+      return {
+        pathname: '',
+        search: '',
+        hash: '',
+        state: undefined,
+      };
+    }
+  }
+}
+
+/**
+ * Given a main or secondary location returns a location that has both of them in single URL. Inverse of
+ * getPartialLocation().
+ * @param newPartialLocation Main or secondary location that should changed
+ * @param currentFullLocation Current full location, so both main and secondary location inside a query param
+ * @param param The name of the param for secondary location
+ * @param isMain
+ */
+function partialLocationToFull(
+  newPartialLocation: H.Location,
+  currentFullLocation: H.Location,
+  param: string,
+  isMain: boolean
+) {
+  const newPartialLocationCopy = { ...newPartialLocation };
+  const currentFullLocationCopy = { ...currentFullLocation };
+  if (isMain) {
+    // get current secondary location, which we are not changing but we need to add it to the new main location as
+    // search param.
+    const secondaryLocation = getPartialLocation(currentFullLocationCopy, param, false);
+    if (secondaryLocation.pathname !== '') {
+      const paramsCopy = new URLSearchParams(newPartialLocationCopy.search);
+      paramsCopy.set(param, secondaryLocation.pathname + secondaryLocation.search);
+      newPartialLocationCopy.search = paramsCopy.toString();
+    }
+    return newPartialLocationCopy;
+  } else {
+    // we keep the current location as it is and just change the secondary location in the search param
+    const paramsCopy = new URLSearchParams(currentFullLocationCopy.search);
+    if (newPartialLocationCopy.pathname === '') {
+      paramsCopy.delete(param);
+    } else {
+      paramsCopy.set(param, newPartialLocationCopy.pathname + newPartialLocationCopy.search);
+    }
+    currentFullLocationCopy.search = paramsCopy.toString();
+    return currentFullLocationCopy;
+  }
+}
+
+/**
+ * Because history api can accept either simple string location or an object we normalize the string one into an object
+ * so we can consider only one type of location later on.
+ * @param location
+ */
+function normalizeLocation(location: H.Path | H.LocationDescriptor<unknown>): H.Location {
+  if (typeof location === 'string') {
+    const url = new URL('http://grafana' + location);
```

Comment:
Do we need to handle cases where grafana is served from an appSubUrl?

Reply:
I use the `new URL` just to parse the location string and then use the pathname, search etc from it. The domain and protocol is required for parsing but it does not really matter what it is here. Will probably put it in some string constant with a comment.